src/data/nat/totient.lean

```diff
@@ -288,4 +288,45 @@ end
 @[simp] lemma totient_two : φ 2 = 1 :=
 (totient_prime prime_two).trans (by norm_num)
 
+/-- An alternative statement of Euler's product formula for the totient function -/
+theorem totient_mul_prod_factors (n : ℕ) :
+  φ n * ∏ p in n.factors.to_finset, p = n * ∏ p in n.factors.to_finset, (p - 1) :=
+begin
+  by_cases hn : n = 0, { simp [hn] },
+  rw multiplicative_factorization φ @totient_mul totient_one hn,
+  simp only [←factorization_finset_prod, ←finsupp.prod_mul],
+  nth_rewrite 1 ←factorization_prod_pow_eq_self hn,
+  rw ←finsupp.prod_mul,
+  refine finsupp.prod_congr (λ p hp, _),
+  have h := zero_lt_iff.mpr (finsupp.mem_support_iff.mp hp),
+  rw [totient_prime_pow (prime_of_mem_factorization hp) h,
+    mul_comm, ←mul_assoc, ←pow_succ, nat.sub_add_cancel h],
+end
```

Comment:
I think it would be useful to have one more additional form for the totient function, which can be factored out of this proof.
```lean
theorem totient_mul_prod_factors' {n : ℕ} (hn : n ≠ 0) :
  φ n = n.factorization.prod (λ p k, p ^ (k - 1) * (p - 1)) :=
begin
  rw multiplicative_factorization φ @totient_mul totient_one hn,
  apply finsupp.prod_congr (λ p hp, _),
  have h := zero_lt_iff.mpr (finsupp.mem_support_iff.mp hp),
  rw [totient_prime_pow (prime_of_mem_factorization hp) h],
end

theorem totient_mul_prod_factors (n : ℕ) :
  φ n * ∏ p in n.factors.to_finset, p = n * ∏ p in n.factors.to_finset, (p - 1) :=
begin
  by_cases hn : n = 0, { simp [hn] },
  rw totient_mul_prod_factors' hn,
  nth_rewrite 2 ←factorization_prod_pow_eq_self hn,
  simp only [←factorization_finset_prod, ←finsupp.prod_mul],
  refine finsupp.prod_congr (λ p hp, _),
  rw [finsupp.mem_support_iff, ← zero_lt_iff] at hp,
  rw [mul_comm, ←mul_assoc, ←pow_succ, nat.sub_add_cancel hp],
end
```